ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc

```diff
@@ -118,15 +119,116 @@ int64 GetTransferredBytes(const NT& node) {
+  } else if (std::is_same<N, Execution>::value) {
+    GetExecutionTypesByIDRequest request;
+    request.add_type_ids(selected_node.type_id());
+    GetExecutionTypesByIDResponse response;
+    TF_RETURN_IF_ERROR(store.GetExecutionTypesByID(request, &response));
+    update_type = response.execution_types(0);
+  } else {
+    GetContextTypesByIDRequest request;
+    request.add_type_ids(selected_node.type_id());
+    GetContextTypesByIDResponse response;
+    TF_RETURN_IF_ERROR(store.GetContextTypesByID(request, &response));
+    update_type = response.context_types(0);
+  }
+  return tensorflow::Status::OK();
+}
+
+template <typename N>
+tensorflow::Status InsertMakeUpNodeIntoDb(MetadataStore& store, N& node) {
+  if (std::is_same<N, Artifact>::value) {
+    PutArtifactsRequest request;
+    request.add_artifacts()->CopyFrom(node);
+    PutArtifactsResponse response;
+    TF_RETURN_IF_ERROR(store.PutArtifacts(request, &response));
+    node.set_id(response.artifact_ids(0));
+  } else if (std::is_same<N, Execution>::value) {
+    PutExecutionsRequest request;
+    request.add_executions()->CopyFrom(node);
+    PutExecutionsResponse response;
+    TF_RETURN_IF_ERROR(store.PutExecutions(request, &response));
+    node.set_id(response.execution_ids(0));
+  } else {
+    PutContextsRequest request;
+    request.add_contexts()->CopyFrom(node);
+    PutContextsResponse response;
+    TF_RETURN_IF_ERROR(store.PutContexts(request, &response));
+    node.set_id(response.context_ids(0));
+  }
+  return tensorflow::Status::OK();
+}
+
+template <typename T, typename N>
+tensorflow::Status PrepareNodeForUpdate(const T& insert_type, const int64 i,
+                                        MetadataStore& store,
+                                        std::vector<Node>& existing_nodes,
+                                        N& node, Type& update_type) {
+  if (!existing_nodes.empty()) {
+    Node existing_node = existing_nodes.back();
+    existing_nodes.pop_back();
+    node = absl::get<N>(existing_node);
+    // node.set_type_id(absl::get<N>(existing_node).type_id());
+    // node.set_name(absl::get<N>(existing_node).name());
+    TF_RETURN_IF_ERROR(
+        SetTypeForUpdateNode(absl::get<N>(existing_node), store, update_type));
+  } else {
+    update_type = insert_type;
+    node.set_type_id(insert_type.id());
+    node.set_name(
+        absl::StrCat("makeup_node", absl::FormatTime(absl::Now()), "_", i));
+    TF_RETURN_IF_ERROR(InsertMakeUpNodeIntoDb(store, node));
+  }
+  return tensorflow::Status::OK();
+}
+
+template <typename T, typename N>
+int64 SetNodePropertiesGivenType(const NodesParam& nodes_param, const T& type,
+                                 N& node) {
+  // Uses "********" as the fake property value for current node.
+  std::string property_value(nodes_param.string_value_bytes, '*');
+  // Loops over the types properties while generating the node's properties
+  // accordingly.
+  // TODO(briansong) Adds more property types support.
```

Comment:
```
CP = MIN(|custom_properties|, nodes_param.num_properties) 

+ node.custom_properties clean |CP| 

RPC = remain_properties_to_change = (nodes_param.num_properties - CP)

```

1. for insert, no-op
2. for update, 
a) RPC = 0, return
b) RPC > 0, the same as the insert case

```
// for existing custom_properties, deleted entry
+ node.custom_properties clean |CP| 

if (rpc == 0) return;

// for existing properties, updated entry
for (const auto& p: type.properties()) {
    (*node..... ) = new value
    if (--rpc == 0) break;
}

// for custom properties, inserted new entry
while (rpc-- >= 0) {
   (*node...) custom properties = new value
}
```